hooks/pre/20-certs_update.rb

```diff
@@ -0,0 +1,66 @@
+require 'fileutils'
+require 'English'
+
+if module_enabled?('katello')
+  SSL_BUILD_DIR = param('certs', 'ssl_build_dir').value
+  CHECK_SCRIPT = `which katello-certs-check`.strip
+  
+  def mark_for_update(cert_name, hostname = nil)
+    path = File.join(*[SSL_BUILD_DIR, hostname, cert_name].compact)
+    if app_value(:noop)
+      puts "Marking certificate #{path} for update (noop)"
+    else
+      puts "Marking certificate #{path} for update"
+      FileUtils.touch("#{path}.update")
+    end
+  end
+
+  ca_file   = param('certs', 'server_ca_cert').value
+  cert_file = param('certs', 'server_cert').value
+  key_file  = param('certs', 'server_key').value
+  
+  if app_value('certs_update_server_ca') && !module_enabled?('katello')
+    fail_and_exit("--certs-update-server-ca needs to be used with katello", 101)
+  end
+  
+  if param('foreman_proxy_certs', 'foreman_proxy_fqdn')
+    hostname = param('foreman_proxy_certs', 'foreman_proxy_fqdn').value
+  else
+    hostname = param('certs', 'node_fqdn').value
+  end
+  
+  if app_value('certs_update_server')
+    mark_for_update("#{hostname}-apache", hostname)
+    mark_for_update("#{hostname}-foreman-proxy", hostname)
+  end
+  
+  if app_value('certs_update_all') || app_value('certs_update_default_ca') || app_value('certs_reset')
+    all_cert_names = Dir.glob(File.join(SSL_BUILD_DIR, hostname, '*.noarch.rpm')).map do |rpm|
+      File.basename(rpm).sub(/-1\.0-\d+\.noarch\.rpm/, '')
+    end.uniq
+  
+    all_cert_names.each do |cert_name|
+      mark_for_update(cert_name, hostname)
+    end
+  end
+  
+  if app_value('certs_update_server_ca') || app_value('certs_reset')
+    mark_for_update('katello-server-ca')
+  end
+  
+  if !app_value('certs_skip_check') &&
+     cert_file.to_s != "" &&
+     (app_value('certs_update_server_ca') || app_value('certs_update_server'))
+    check_cmd = %(#{CHECK_SCRIPT} -c "#{cert_file}" -k "#{key_file}" -b "#{ca_file}")
+    output = `#{check_cmd} 2>&1`
+    unless $CHILD_STATUS.success?
+      fail_and_exit("Command '#{check_cmd}' exited with #{$CHILD_STATUS.exitstatus}:\n #{output}", 101)
+    end
+  end
+  
+  if app_value('certs_reset') && !app_value(:noop)
+    param('certs', 'server_cert').unset_value
+    param('certs', 'server_key').unset_value
+    param('certs', 'server_ca_cert').unset_value
+  end
```

Comment:
I also wonder how this worked. In `pre` all parameter values are stored in the answers file. Here it then resets them. Doesn't sound like that will persist between runs.

Reply:
My understanding is that `--certs-reset` is intended to restore to the default paths -- it's a way of starting over.

The user would first run the installer with `--certs-reset` to restore to the katello defaults. Then if defaults aren't their end goal and they need to supply new certificates they would run the installer again and supply the new answers.

Do you think it should work differently?

Reply:
Reading Kafo, it suggests that you are not guaranteed that it will commit the changes to disk this way. From the [Kafo Hooks docs](https://github.com/theforeman/kafo#hooks):
> pre - just before puppet is executed to converge system, after parameter values are stored in the answer file

My interpretation is that you should not change any parameters in the `pre` stage, but really before that.

From reading those same docs, `pre_values` makes the most sense to me. This allows you to use `foreman-installer --certs-reset --interactive` which currently is not possible.

Another argument is that this now doesn't work together with the `pre_commit` hook this change has. You run `katello-certs-check` on some values and then reset the values.

Now this is only from reading docs so some real tests would be good to see if the theory matches reality :)

Reply:
Indeed, so I tested this on Katello 3.16 with the below approach:

1. Made a backup of my answers file, then edited answers to include:

```yaml
certs:
  server_cert: "/root/no-such-cert.pem"
  server_key: "/root/no-such-key.pem"
  server_ca_cert: "/root/no-such-ca-cert.pem"
```

2. For now I commented lines related to required CPU/memory check since I am running these tests on a Forklifted box.

3. # foreman-installer --certs-reset

Then I checked the answers file and indeed the values I set in step 1 were persisted. I agree this doesn't seem correct. I'll experiment with moving it to another stage as suggested. Looking at some other examples we may also need `kafo.config.configure_application`

Reply:
I think the correct approach would be to move `unset_value` to `pre_commit` as I suggested. `pre` means `post_commit` and `pre_puppet_apply`. Using `kafo.config.configure_application` would mean saving things to disk again, which is unexpected.